[WIP] Multiple Imputation: Example with IterativeImputer

RianneSchouten opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
As promised in #11259 and following on @sergeyf 's work in #11314 and #11350: in this PR an example that shows how to use IterativeImputer for Multiple Imputation.

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

#### What does this implement/fix? Explain your changes.

As discussed in #11259, the defaults of IterativeImputer are such that single imputation is performed. Because the method is also quite powerfull for Multiple Imputation, we agreed to make an example that shows the user how to use ImputerImputer to perform Multiple Imputation.

I made the document: examples/plot_multiple_imputation.py and it shows 2 things:
1. Estimation of beta estimates and their standard error: compare IterativeImputer with using IterativeImputer as a MICE Imputer. 
2. How to use IterativeImputer as a MICE Imputer when making a prediction model (with train and test datasets).

The following two figures are created with the script:
![plotmultipleimputationstatisticalinference](https://user-images.githubusercontent.com/22293115/41985588-041d55f0-7a34-11e8-9352-160686c08eaa.png)
![plotmultipleimputationprediction](https://user-images.githubusercontent.com/22293115/41985591-07b072b0-7a34-11e8-9158-3fbafdf977ba.png)

#### Any other comments?

The script use the IterativeImputer as it is currently accepted in the master branch (with name: ChainedImputer). When PR #11350 is accepted, I will update the code with the newest arguments (f.e. no n_burn_in but just n_iter). 

Are there suggestions for improvements? 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

RianneSchouten committed:
add example multiple imputation

RianneSchouten committed:
adjust figure widths and legends

sklearn-lgtm commented:
This pull request **introduces 2 alerts** when merging 965ae8ef42b6870a85eaa3f596a266fdca424ae0 into 3b5abf76597ce6aff76192869f92647c1b5259e7 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-7fc582af033b9efc4ecbb1674ab394ee43d3db16)

**new alerts:**

* 1 for Unused import
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

sergeyf commented:
Thanks @RianneSchouten!

Take a look at http://scikit-learn.org/stable/developers/contributing.html#contributing-pull-requests for info about fixing the flake errors. `sklearn` has some strict style requirements =)

stefvanbuuren commented:
Great. Two remarks. 

1. You say: "The strength of the method is that it allows for finding unbiased statistical estimates due to its chained character."

This is a broad statement, and only true if the imputations are drawn from the posterior.

2. I am not familiar with Python's style, but perhaps some interpretation of the results would be useful. Your show the methods produce - as I expected - different standard errors. Is smaller better? How do we know which estimate is correct?

jnothman commented:
Here is the rendered documentation. I hope to look over it soon.

jnothman commented:
https://26355-843222-gh.circle-artifacts.com/0/doc/auto_examples/plot_multiple_imputation.html

jnothman commented:
In the first instance, you'll need to make those plots bigger...

RianneSchouten committed:
changed code according pep rules and increased figure size

RianneSchouten committed:
solve two issues from lgtm and improve introduction text

RianneSchouten commented:
@sergeyf Thanks for the tip, I read the documents and think I changed the code according to the rules.
@stefvanbuuren  I have changed the text into the following to listen to your remark 1. Do you think this is better?

`The chained character of the method and the possiblity to draw imputation values from the posterior
distribution of a Bayesian imputation model allows for the finding of unbiased statistical estimates.`

Regarding remark 2: I can add some explanation of course, but I am not sure where to place that. @sergeyf , @jnothman is there a place I can write some explanation/interpretation of the images. 
@jnothman : I made the images twice as large. Is that okay?

jnothman reviewed (commented):
In terms of where you can make comments/explanation/interpretation of plots: historically this has all been in the passage at the top. You can, however, have integrated text blocks with comments, as in examples/preprocessing/plot_all_scaling.py.

RianneSchouten committed:
remove spaces in arguments and add lines for definitions

jnothman commented:
The example took more than 10 minutes to run on Circle CI, causing it to halt. I don't know if this is a bug, but we need our examples to run a lot faster, even if that makes them less real-world.

jnothman commented on examples/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,472 @@
+    if mech == "MNAR":
+        for i in np.arange(n_features):
+            data_values = -np.mean(X[:, i]) + X[:, i]
+            weights = list(map(lambda x: math.exp(x) / (1 + math.exp(x)),
```
why is this not just `np.exp(data_values) / (1 + np.exp(data_values))` (or `1 / (1 + np.exp(-data_values))`)?

RianneSchouten replied:
Think you're absolutely right. Changed it.

jnothman reviewed (commented):
I'm separately interested in your "amputation functions". We had lots of discussion about putting this kind of feature into the library at #7084, but we were not aware of a literature in this space, and were disappointed by some of the over-simplifications in that PR.

For MCAR, you seem to take an approach which more likely masks high values. Looks neat enough. In reality should the base of the exponential be a parameter to allow the removal to be more/less skewed to the upper extreme? Are there other more interesting MCAR amputators you would recommend? Perhaps comment at #6284.

RianneSchouten commented:
@jnothman I'm happy you mention the amputation function: I (we) developed an amputation procedure that generates sophisticated missing data, in different forms, and we put it in the mice package in r (function: ampute). You can read about the procedure: https://rianneschouten.github.io/mice_ampute/vignette/ampute.html . I just got the confirmation that my article about the procedure will be published in Journal of Statistical Computation and Simulation. Will put the link to the online version here next week. 

Apart from this article, no one has written about the generation of multivariate missingness (except Jaap Brand in 1999). The generation of univariate missingness is often done (deleting missing values feature by feature), but it is almost impossible to use the univariate approach to create a good overall missingness structure. Thats why we developed the multivariate procedure in R.

It's been on my wish list for a few months now to have the same procedure available in Python, also because I need it for many of my simulation studies. The effect of missingness mechanisms on prediction models is not yet clear. 

Note that I didnot implement anything like the multivariate amputation procedure in this PR. Both this MCAR and MNAR approach are univariate approaches (and you're right: I create MNAR RIGHT missingness). Apart from RIGHT there are a lot of other types. 
It is with MAR missingness where the multivariate approach is especially useful. 

I will read the PR's you mention tomorrow. 
Maybe I could take some days of from work and write the multivariate amputation procedure in Python code on short notice if that would be useful?

RianneSchouten commented:
@jnothman I put a simulation aspect in the example. I will get it out tomorrow and then it should pass the time test.

jnothman commented:
The vignette (I've only read section 1) looks great! We were onto a similar
track in the univariate case, and having a weighted sum of features makes a
lot of sense... I've not yet understood which feature you mask if a
particular weighted sum is selected for masking.... but I'll  keep reading
eventually.​

RianneSchouten commented:
> I've not yet understood which feature you mask if a
particular weighted sum is selected for masking

That has to do with the missingness patterns. You define patterns beforehand. Lets say you have a dataset with y1, y2 and y3. a pattern might look like: 0 0 1, meaning, if a weighted sum is selected for masking, features y1 and y2 will become incomplete. 

You can define different kinds of patterns, ending up with a dataset with different patterns as well. each missing data pattern is connected to one way of calculating the weighted sum scores. In other words, you define the weights for each pattern.

stefvanbuuren commented:
@RianneSchouten Text is fine.

stefvanbuuren commented:
Both examples are fairly advanced. At some point you might wish to add a minimal example, e.g., as in https://github.com/stefvanbuuren/mice to emphasise the basics.

RianneSchouten committed:
put rules in separate functions and include explanation

RianneSchouten commented:
Okay, I made the explanation at the beginning larger, with more explanation of rubin's rules. also I put the rules in separate functions such that it is clear where the magic happens. In addition, I turned the nr of simulations down from 10 to 2, now the time range should be fine. Let's see. Otherwise I will get the simulation out. 

@stefvanbuuren you might be right that the examples are high level, comparing different approaches while showing the example. 

Do others think I should make a more basic example as well?

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging ed1db8bb6d5e33c50e354de2246324d7c7f19210 into 3b5abf76597ce6aff76192869f92647c1b5259e7 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-798f419d02380f3dfc1984e4a9261a4239db6a4b)

**new alerts:**

* 1 for Encoding error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

RianneSchouten committed:
line from 80 to 79 characters

sklearn-lgtm commented:
This pull request **introduces 1 alert** when merging 7c3fb9705a21f34bd392361233111ea6976054d8 into 3b5abf76597ce6aff76192869f92647c1b5259e7 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-0de86ee43c82ccbd85ae6e2c6852195d7757d1b2)

**new alerts:**

* 1 for Encoding error

---

*Comment posted by [LGTM.com](https://lgtm.com)*

jnothman commented on examples/plot_multiple_imputation.py:
```diff
@@ -0,0 +1,509 @@
+robust summary measures such as medians or ranges instead of using Rubin’s
+pooling rules. This applies to an estimate like explained variance.
+
+In sum, Rubin’s pooling rules are as follows. The overall point estimate after
```
lgtm.com is complaining about this curly apostrophe. Use '.

qinhanmin2014 committed:
DOC Format in DBSCAN

adrinjalali committed:
MNT Change max_bound -> max_eps in OPTICS (#11984)

jnothman committed:
COSMIT remove unnecessary _TreeNode methods (#11983)

adrinjalali committed:
DOC OPTICS: improve docstring and add default values. (#11987)

niedakh committed:
DOC adding scikit-multilearn to related projects list (#11988)

wdevazelhes committed:
TST FIX use match rather than message in pytest.raises (#12001)

Previously these assertions would pass without matching.

jnothman committed:
DOC note controversy on multiclass balanced accuracy definition (#11994)

jakirkham committed:
MNT Use `fmax` when finding the maximum (#12005)

Instead of adding an `if` to check for values that become the new max,
simply use `fmax` to get the maximum and update the value. This improves
readability. It may improve performance as `fmax` can be a single
assembly instruction. Though most compilers can probably figure this out
anyways.

albertcthomas committed:
DOC small changes in outlier detection documentation (#12003)

qinhanmin2014 committed:
MNT Remove n_clusters_ in OPTICS (#11981)

ogrisel committed:
Joblib 0.12.4 (#12007)

This should fix #11971 (fixed PyPy support, pypy3 is now part of the joblib build matrix on travis).

It should also be backported to 0.20.X.

ogrisel committed:
MAINT make pytest collection ignore folders with Python scripts (#12011)

Some IDEs such as VS Code use the pytest command to collect all the tests of
the workspace in the background. This can cause unexpected execution of
arbitrary Python scripts in the workspace (examples, benchmarks...).

The doc folder is also ignored because it has python scripts for sphinx
along with copies of the examples.

To safely run pytest in the doc folder, we need to used the find command
to find all "*.rst" files as done in the project Makefile.

adrinjalali committed:
[MRG+1] break the tie in Meanshift in case cluster intensities are the same (#11901)

gabrielecalvo committed:
DOC fix minor spacing issue in the iris dataset description (#12019)

jnothman committed:
TST use urlopen monkeypatch for test_decode_* (#12020)

Avoid requiring internet for test suite. Examples will still run with internet (as long as cache is occasionally cleared).

vivekk0903 committed:
DOC gradient boosting fit() supports sparse X (#12022)

wdevazelhes committed:
DOC: Add pytest version in documentation (#12002)

papaonlegs committed:
DOC fix for linnerud dataset (#12024)

The descriptions were the wrong way around

ogrisel committed:
MAINT skip joblib vendor test on debian (#12027)

jeremiedbb committed:
MNT Fix utils.sparse import in neural_network.rbm (#12032)

jnothman committed:
MNT Revert the deprecation of min_samples_leaf and min_weight_fraction_leaf (#11998)

adrinjalali committed:
OPTICS remove redundant recursion (#11985)

maxcopeland committed:
EXA use openml fetcher in plot_gpr_co2.py  example (#12004)

rth committed:
CI Workaround to test numpy 1.8.2 and scipy 0.13.3 (#12042)

jnothman committed:
FIX ordering_ type and cosmetic changes to structure for OPTICS main loop (#11986)

jnothman committed:
[MRG] MNT rename min_cluster_size_ratio to min_cluster_size (#11913)

Commit:
DOC `sample_weight` removed from the docs in `SVR` class. (#12046)

JarnoRFB committed:
MNT Unify and refactor strategy error (#12050)

adrinjalali committed:
[MRG] DOC Examples added to the rest of linear models (#11975)

TomDLT committed:
DOC Generated author list from github (#11708)

JarnoRFB committed:
ENH Allow scoring of dummies without testsamples (#11957)

zjpoh committed:
DOC Fix docstring inconsistency in nmf.py (#12063)

adrinjalali committed:
MAINT Fix invalid escape sequence (#12064)

zdgriffith committed:
DOC fix typos in documentation. (#12059)

qinhanmin2014 committed:
DOC Include fetch_openml doc in user guide (#12065)

rth committed:
MNT: Anonimize IP for Google Analytics (#12038)

rth committed:
TST Use pytest.raises instead of legacy constructions (#12057)

jorisvandenbossche committed:
BUG always raise on NaN in OneHotEncoder for object dtype data (#12033)

ogrisel committed:
MAINT joblib 0.12.5 (#12066)

thomasjpfan committed:
ENH Adds drop in FeatureUnion (#11640)

zjpoh committed:
ENH Better error message for metrics of neighbors (#11914)

qinhanmin2014 committed:
MNT Duplicate import in test_pipeline.py

Travis failed to detect it in #11640

zjpoh committed:
ENH Better error message for sparse metrics of neighbors (#12073)

glemaitre committed:
MAINT: skip doctest for Python 2 (#12074)

amueller committed:
fix typo

adrinjalali committed:
FIX OPTICS Change quick_scan floating point comparison to isclose (#11929)

glemaitre committed:
TST: skip test requiring internet using --skip-network (#12067)

zdgriffith committed:
DOC Fix description of SVC intercept_ shape in user guide (#12070)

massich committed:
TST Ignore warnings in common test to avoid collection errors (#12093)

adrinjalali committed:
DOC Fix optics metric issues (DOC and precomputed) (#12028)

jnothman committed:
DOC move OPTICS to 0.21

kno10 committed:
DOC Reword to avoid that people draw wrong conclusions (#12095)

rth committed:
MNT Only checks warnings on latest depedendencies versions in CI (#12048)

rth committed:
[MRG] MNT Re-enable PyPy CI (#12039)

adrinjalali committed:
DOC BaggingRegressor missing default value for oob_score in docstring (#12108)

louib committed:
DOC Removing quotes from variant names. (#12113)

oliverrausch committed:
[MRG] Update test_metaestimators to pass y parameter when calling score (#12089)

albertcthomas committed:
DOC Update fit_transform docstring of OneHotEncoder (#12117)

qinhanmin2014 committed:
DOC Typo in OneHotEncoder

Mottl committed:
Fix typo (#12126)

Fix typo in `examples/compose/plot_column_transformer.py`

jorisvandenbossche committed:
[MRG +1] ColumnTransformer: store evaluated function column specifier during fit (#12107)

TomDLT committed:
ENH (0.21) Make OPTICS more memory efficient when calling kneighbors (#12103)

Sriharsha-hatwar committed:
Added the changes to remove the documentation support statements for Python 2 (#12083)

Sriharsha-hatwar committed:
DOC Replaced the deprecated early_stopping parameter with n_iter_no_change. (#12133)

adrinjalali committed:
[MRG] DOC covariance doctest examples (#12124)

qinhanmin2014 committed:
DOC (0.21) OPTICS Note the order of reachability_ and core_distances_ (#12132)

NicolasHug committed:
DOC Removed duplicated doc in tree.rst (#11922)

ogrisel committed:
[MRG] Fix FutureWarnings in logistic regression examples (#12114)

* Be more specific about logistic regression solver in examples

* Use early stopped SGD (faster) and plot cross-validated error for best models

* Fix LR solver in /plot_voting_probas.pyexamples/ensemble/plot_voting_probas.py

* Fix LR solver & scale data in plot_digits_classification_exercise.py

* Use saga solver in plot_logistic_l1_l2_sparsity.py

* Use LBFGS solver in plot_iris_logistic.py

* Use LBFGS in plot_logistic.py

* Use SAGA solver for Logistic Regression Path example

* Use LBFGS solver in plot_classifier_chain_yeast.py

* Use LBFGS solver in plot_rbm_logistic_classification.py

* typo

* typo

* Bump up pandas dependency to 0.17.1

* Bump up examples minimal deps to match pandas 0.17.1

* Fix figure layout for plot_digits_pipe.py

* Version numbers are not decimal numbers

* Set multinomial, no scaling to keep example simple, fix formatting of example doc

* Missing plt.tight_layout() in plot_voting_probas.py

* Missing plt.tight_layout() in plot_logistic.py

glemaitre committed:
BUG: check equality instead of identity in check_cv (#12155)

vinayak-mehta committed:
[MRG] Convert ColumnTransformer input list to numpy array (#12104)

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #12096.

#### What does this implement/fix? Explain your changes.
Converts the input list for ColumnTransformer to a numpy array.

Added a check inside `transform` and `fit_transform` to check if the input `X` is a list, if it is then it gets converted to a numpy array.

#### Any other comments?
Should this conversion be documented in the docstrings for ColumnTransfomer's `fit`, `transform` and `fit_transform`?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

datajanko committed:
add sparse_threshold to make_columns_transformer (#12152)

fixes issue #12149

janvanrijn committed:
ColumnTransformer generalization to work on empty lists (#12084)

jorisvandenbossche committed:
TST update make_column_transformer test + add comment (#12156)

Follow-up on https://github.com/scikit-learn/scikit-learn/pull/12152
And added comment why transformer_weights is not passed through, see
https://github.com/scikit-learn/scikit-learn/pull/11183#pullrequestreview-125539051
for more discussion

djsutherland committed:
coef0 is a float, not an int (#12161)

jorisvandenbossche committed:
DOC: update link to other versions (#12158)

amueller committed:
DOC added 0.20 contributors to whats new. (#12160)

qinhanmin2014 committed:
DOC More specific about the limitation of make_column_transformer (#12163)

jnothman committed:
MAINT update comment

qinhanmin2014 committed:
MNT Avoid using "is" when comparing strings (#12168)

ogrisel committed:
OPTIM remove useless overhead caused by nested parallelism in mean_shift (#12159)

qinhanmin2014 committed:
MNT Unused import in plot_gpr_co2.py

ogrisel committed:
[MRG] Crash when using SGDClassifier with early stopping in a parallel grid search (#12122)

ogrisel committed:
Dedicate the release to Raghav

ogrisel committed:
DOC start section for the 0.20.1 bugfix notes (#12170)

jnothman committed:
DOC add a known issue entry for euclidean_distances precision (#12176)

tomMoral committed:
Fix parallel backend neighbors (#12172)

TomDLT committed:
[MRG] improve check_non_negative for sparse matrices (#12106)

jnothman committed:
FIX Use take instead of choose in compute_sample_weight (#12165)

qinhanmin2014 committed:
DOC Add sections to whats new 0.20.1 (#12183)

haroldfox committed:
DOC KDE normalisation clarified (#11275)

TomDLT committed:
[MRG] Fix diagonal in DBSCAN with precomputed sparse neighbors graph (#12105)

qinhanmin2014 committed:
MNT Move what's new entry

J0 committed:
DOC Add Versionadded tag to sklearn/_config.py (#12187)

adrinjalali committed:
BaseSearchCV._run_search raises NotImplementedError instead of being an abstractmethod (#12182)

* _run_search raises NotImplementedError instead of being and abstractmethod

* add error message

* test for a BaseSearchCV child w/o a _run_search

* make the test python2 compatible, still in 0.20 zone.

* specify cv in tests not to trigger the related FutureWarning

* PEP8

qinhanmin2014 committed:
DOC Add versionadded to set_config (#12196)

LilyX2021 committed:
DOC Improve ColumnTransformer docstrings (#12206)

rebekahkim committed:
MNT Remove duplicate import of warnings & unused variables (#12203)

Commit:
DOC Fix typo in neighbors/nearest_centroid.py (#12223)

markhannel committed:
DOC Fixing summary table in the linear model documentation. (#12220)

qinhanmin2014 committed:
MNT Use name instead of float to specify colors (#12199)

ygivenx committed:
[MRG] More informative error message in OneHotEncoder(categories=None) with negative integer values (#12180)

* Fix Issue #12179

OneHotEncoder "only non-negative integers" message should suggest using
categories='auto'

* Fix Issue #12179

    OneHotEncoder "only non-negative integers" message should suggest using
    categories='auto'

* Fix Issue #12179

OneHotEncoder "only non-negative integers" message should suggest using
categories='auto'

* Fixes #12180 Modify the error message

* Fix the spacing

amueller committed:
add explicit mention of scaing for saga in logisticregression docs. (#12236)

samwaterbury committed:
MNT Raise error for duplicate classes when constructing a MultiLabelBinarizer (#12195)

qinhanmin2014 committed:
DOC Encourage contributors to use sklearn.show_versions() (#12225)

qinhanmin2014 committed:
MNT Add versionadded to set_config parameters

qinhanmin2014 committed:
MNT Unused imports in examples

VarIr committed:
Fix numpy.int overflow in make_classification (#10811)

qinhanmin2014 committed:
MNT Remove duplicate entry in whats new

rth committed:
Remove test_import_sklearn_no_warnings (#12244)

rasbt committed:
DOC fix cross-entropy typo in tree docs (#12242)

Commit:
[MRG] Added Tips in SVM user guide for tuning C parameter in LinearSVC and LinearSVR (#12185)

* Added Tip for tuning C parameter in LinearSVC/SVR

* Added reference about performance improvement

* Clearer explanation about large C values

* Update svm.rst

rth committed:
DOC what's new entry for "Fix numpy.int overflow in make_classification #10811"

TomDLT committed:
[MRG] Fast PolynomialFeatures on dense arrays (#12251)

amueller committed:
[MRG + 1] return_train_score deprecation (#12241)

rth committed:
Remove unused private functions (#12253)

jacquerie committed:
DOC Remove mentions of removed AUTHORS.rst file (#12262)

qinhanmin2014 committed:
MNT Change show_versions format to suit markdown (#12255)

jnothman committed:
DOC add note on discretization creating non-linearity (#12269)

TakingItCasual committed:
MNT Updated PyPI URLs (#12274)

amueller committed:
MNT complete VotingClassifier flatten_transform deprecation (#12252)

jnothman committed:
DOC Move 'for instance' to front

adrinjalali committed:
[MRG+1] ColumnTransformer fix having mixed types in a single passthrough (#12200)

* fix having mixed types in a single passthrough

* use check_array

* add comment

* Reference the relevant issue in the test

* don't force finite on check_array

* add the whats_new entry

* modify whats_new entry for a [probably] better description.

* raise a value error if columns can't be stacked.

* test the custom ValueError

* take sparse.hstack out of try/catch

* improve whats_new entry

* add comment for the dtype conversion

* change the tests to use an np of dtype 'O' instead

NicolasHug committed:
[MRG] Created 'cross-validation estimator' entry in glossary (#11661)

* Created 'cross-validation estimator' in glossary and referenced it where
needed

* Addressed adrinjalali comments

* updated glossary entry

* updaed docstrings according to commets

sjtrny committed:
DOC removed ambiguity in pipeline gridsearch example (#12272)

* Removed ambiguity in pipeline gridsearch example

* Formatted to pep8

TakingItCasual committed:
MNT Converting http to https (#12277)

qinhanmin2014 committed:
MNT Move what's new entry

TakingItCasual committed:
Converting http to https (2)... (#12292)

andreanr committed:
DOC Add class example for LedoitWolf (#12214)

janvanrijn committed:
FIX cache of OpenML fetcher (#12246)

benjastudio committed:
EXA Fix bad data visualisation in "Importance of Feature Scaling" (#12280)

jacquerie committed:
MNT Make check_X_y raise a better error when y is None (#12283)

gauravahlawat01goal committed:
DOC Added "mars" testimonial to testimonials page (#12298)

qinhanmin2014 committed:
DOC What's new typo

kushalchauhan98 committed:
DOC Fix broken link to joblib documentation (#12301)

kataev committed:
MNT Apply pep8 to docs code (#12275)

TakingItCasual committed:
Converting http to https (3)... (#12302)

thomasjpfan committed:
MNT Refactors doc test into seperate script (#12248)

kno10 committed:
ENH (0.21) Remember predecessor in OPTICS (#12135)

This will allow filtering the extracted clusters,
in order to improve quality (see Schubert and Gertz 2018)

adrinjalali committed:
FIX (0.21) OPTICS correctly handle multiple infs in reachability array. (#12029)

oleksandr-pavlyk committed:
TST Use same random seeds for both GaussianMixture.fit (#12307)

kiote committed:
ENH Cache class mapping in MultiLabelBinarizer() (#12116)

Fixes #11680

qinhanmin2014 committed:
ENH Raise an error when pos_label is not in binary y_true (#12313)

oleksandr-pavlyk committed:
TST Use v_measure_score to compare label arrays up to permutation (#12265)

qinhanmin2014 committed:
DOC What's new typo

Commit:
EXA Fix title, overlapping plots and axis labels in plot_ols_ridge_variance.py (#12296)

janvanrijn committed:
[MRG] Fix fetch_openml when ignore attributes are numeric (#12330)

* modularized data column functionality

* small bugfix

* removes redundant line breaks

* added some documentation on the added fn

* added additional comment on advice of Nicholas Hug

* added test case

* merged master into branch, and added small comments by Joel

* added doc item

chang committed:
[MRG] FIX Update power_transform docstring and add FutureWarning (#12317)

dvukolov committed:
DOC check_array() and check_X_y() documentation update (#12340)

adrinjalali committed:
ENH add get_n_leaves() and get_max_depth() to DesicionTrees (#12300)

adrinjalali committed:
DOC fix logistic regression.fit docstring on y (#12343)

rohan-varma committed:
[MRG] Move RandomTreesEmbedding criterion & max_features to be class attributes (#12324)

* rm criterion and max_features from __init__ and store them as class attrs instead

* make sure that the docstring comes first

whiletruelearn committed:
ENH (0.21) Add max_error to the existing set of metrics for regression (#12232)

chang committed:
DOC Update v0.20.rst with power_transform API change (#12351)

amueller committed:
MNT simple deprecations and removals for 0.21 (#12238)

Part of #11992.
These were all the things that seemed pretty straight-forward. It's actually a bit bulky but should still be easy to review, hopefully.

amueller committed:
[MRG] Matplotlib tree plotting (#9251)

* add reingold tillford tree layout algorithm

* add first silly implementation of matplotlib based plotting for trees

* object oriented design for export_graphviz so it can be extended

* add class for mlp export

* add colors

* separately scale x and y, add arrowheads, fix strings

* implement max_depth

* don't use alpha for coloring because it makes boxes transparent

* remove unused variables

* vertical center of boxes

* fix/simplify newline trimming

* somewhere in the middle of stuff

trying to get rid of scalex, scaley

* remove "find_longest_child" for now, fix tests

* make scalex and scaley internal, and ax local.

render everything once to get the bbox sizes, then again to actually plot it with known extents.

* add some margin to the max bbox width

* add _BaseTreeExporter baseclass

* add docstring to plot_tree

* use data coordinates so we can put the plot in a subplot, remove some hacks.

* remove scalex, scaley, add automatic font size

* use rendered stuff for setting limits (well nearly there)

* import plot_tree into tree module

* set limits before font size adjustment?

* add tree plotting via matplotlib to iris example and to docs

* pep8 fix

* skip doctest on plot_tree because matplotlib is not installed on all CI machines

* redo everything in axis pixel coordinates

re-introduce scalex, scaley
add max_extents to tree to get tree size before plotting

* fix max-depth

parent node positioning and don't consider deep nodes in layouting

* consider height in fontsize computation

in case someone gave us a very flat figure

* fix error when max_depth is None

* add docstring for tree plotting fontsize

* starting on jnothman's review

* renaming fixes

* whatsnew for tree plotting

* clear axes prior to doing anything.

* fix doctests

* skip matplotlib doctest

* trying to debug circle failure

* trying to show full traceback

* more print debugging

* remove debugging crud

* hack around matplotlib <1.5 issues

* copy bbox args because old matplotlib is weird.

* pep8 fixes

* add explicit boxstyle

* more pep8

* even more pep8

* add comment about matplotlib version requirement

* remove redundant file

* add whatsnew entry that the merge lost

* fix merge issue

* more merge issues

* whitespace ...

* remove doctest skip to see what's happening

* added some simple invariance tests buchheim function

* refactor
___init__ into superclass

* added some tests of plot_tree

* put skip back in, fix typo, fix versionadded number

* remove unused parameters special_characters and parallel_leaves from mpl plotting

* rename tests to test_reingold_tilford

* added license header from pymag-trees repo

* remove duplicate test file.

RianneSchouten committed:
save latest changes

RianneSchouten committed:
Merge branch 'master' into multiple-imputation

RianneSchouten commented:
I wanted to bring my multiple-imputation branch up to date with the master branch of scikit learn where I originally forked from. 

I first brought my master branch up to date by:
git fetch upstream
git checkout master
git merge upstream/master

I then brought the multiple-imputation branch up to date by:
git fetch -p origin
git merge origin/master
git checkout multiple-imputation
git merge master.

Now I see that at the top of this page it says: rianneschouten wants to push 164 commit into scikit-learn:iterativeimputer.

Can someone explain to me if I did something wrong an/or how this works?

RianneSchouten commented:
I made a mess.
Since my work was in only one file, i will start over.

RianneSchouten commented:
As the developer info says, I will fork/clone the master branch, make a local new branch, and work on the ampute files in that branch and then make a pull request. 

I am wondering though, if i want to improve/update the multiple-imputation example, i have to use the info in the IterativeImputer branch. Should I wait till you guys update the master branch with that information or should I fork/clone the IterativeImputer branch? What is the preferred way to do it? 

Thanks in advance.

sergeyf commented:
@RianneSchouten The master will not be updated until we have all the pieces of `IterativeImputer` done in its own branch. That includes examples, etc.

RianneSchouten commented:
That makes sense. Thanks.

RianneSchouten commented:
Hey guys,

I am sorry but it seems I don't have enough time to actively finish this.
The example is finished in the sense: it shows how to loop over the iterativeimputer to create multiple imputations, how to calculate the variance of beta estimates, how to pool the outcomes after multiple imputation.

The example should be finished by: applying the code to the newest version of the iterative imputer, and by removing the simulation aspect from the example (that is not necessary to show how to perform mi and takes unncessary time).

The difficult aspect is that the pooling rules change for the different statistical analysis methods. In the example, I use linear regression which is the most used and most basic analysis method.

I hope someone else can finish this so that it comes along with the iterativeimputer?

Kind regards,
Rianne

RianneSchouten commented:
these pooling rules are good for linear regression. so the example works
and is good.

in case of other statistical models you need other pooling rules.
you could just mention that and then add a reference such as:
Combining estimates of interest in prognostic modelling studies
after multiple imputation: current practice and guidelines
Andrea Marshall*1,2, Douglas G Altman1, Roger L Holder3 and
Patrick Royston4

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virusvrij.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Op wo 12 dec. 2018 om 18:26 schreef Sergey Feldman <notifications@github.com
>:

> Hi @RianneSchouten <https://github.com/RianneSchouten>. When you say "The
> difficult aspect is that the pooling rules change for the different
> statistical analysis methods" - does this mean we need to figure out other
> pooling methods for this particular example? Or is the current pooling
> sufficient to demonstrate how this generally works? Should we include a
> reference for other pooling for other models?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/11370#issuecomment-446671467>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVQqe6ZXGkrqf4wK7unvK9Jv6BUOH7CGks5u4Tw3gaJpZM4U5_br>
> .
>

sergeyf commented:
Hi again @RianneSchouten. Just to confirm what you mean about ` by removing the simulation aspect from the example`. Which lines do you mean I should remove? I assume 445 to 509 (plus doc info about those lines). Is that correct?

sergeyf commented:
Also, @RianneSchouten. This statement: ```Due to the between
imputation variance, the standard errors of all regression coefficients are
larger with multiple imputation than with single imputation. This allows for
valid statistical inference making.```

I think ML people might not know what this means as we don't really use standard errors much. Why do larger standard errors allow for valid statistical inference making?

jnothman commented:
I think the point is that single imputation underestimates uncertainty. So
if you want to know how reliable your prediction is, or what the credible
interval is, you need to account for uncertainty due to missingness.

jnothman commented:
Best to make a new PR and I/others can see what it looks like.

jnothman commented:
Superseded by #13025
